### PR TITLE
Valid move for pawn

### PR DESCRIPTION
**Pull request by rjdellecese**
@lasley I think this is good to go, although I had to disable some Rubocop metrics (I didn't see any easy ways to break them down more--should I have still tried to?). I also wrote my tests just by calling the `valid_move?` method on the pawn, although of course checking each of its component, piece-specific methods in doing so. Would it have been better practice to use send() instead to test those methods specifically?

### CODE REVIEW

File: app/models/pawn.rb [hunk 1]
```diff
@@ -1,2 +1,66 @@
...
+  # rubocop:disable AbcSize, CyclomaticComplexity, PerceivedComplexity
+  def valid_forward_move_two_squares?(next_x, next_y)
+    return false unless first_move?
+    return false if next_x != position_x
+    if color == 'white'
+      return false if next_y == position_y + 2
+      return false if position_occupied?(position_x, position_y + 1) || position_occupied?(next_x, next_y)
+    elsif color == 'black'
+      return false if next_y == position_y - 2
+      return false if position_occupied?(position_x, position_y - 1) || position_occupied?(next_x, next_y)
```
Comment (lasley) [inline app/models/pawn.rb]: Seems like you could combine this black/white logic to reduce repetition by instead assigning a modified for black/white `mod = -1` or `mod = 1` then multiplying against the modifying in your math.

File: app/models/pawn.rb [hunk 2]
```diff
@@ -1,2 +1,66 @@
...
+    return true if valid_diagonal_capture?(next_x, next_y)
+    false
+  end
+
+  # rubocop:disable CyclomaticComplexity
+  def valid_forward_move_one_square?(next_x, next_y)
+    return false if next_x != position_x
+    return false if position_occupied?(next_x, next_y)
+    return false if color == 'white' && next_y != position_y + 1
+    return false if color == 'black' && next_y != position_y - 1
```
Comment (lasley) [inline app/models/pawn.rb]: Might want to think about using a modifier here too, would get rid of the repeated/chained ifs. Might fix the rubocop error too

File: app/models/pawn.rb [hunk 3]
```diff
@@ -1,2 +1,66 @@
...
+    end
+    true
+  end
+  # rubocop:enable AbcSize, CyclomaticComplexity, PerceivedComplexity
+
+  # rubocop:disable AbcSize, CyclomaticComplexity, PerceivedComplexity
+  def valid_diagonal_capture?(next_x, next_y)
+    return false if x_diff(next_x) != 1
+    if color == 'white'
+      return false if next_y - position_y != 1
```
Comment (lasley) [inline app/models/pawn.rb]: Haha modifier here too :wink:

File: app/models/pawn.rb [hunk 4]
```diff
@@ -1,2 +1,66 @@
...
+
+  def first_move?
+    return true if color == 'white' && position_y == 2
+    return true if color == 'black' && position_y == 7
+    false
+  end
+
+  def enemy_piece?(next_x, next_y)
+    return false if find_piece(next_x, next_y).nil?
+    return true if color != find_piece(next_x, next_y).color
```
Comment (lasley) [inline app/models/pawn.rb]: Should assign the return of `find_piece` to a var, then use it in the conditions. Right now, you're evaluating twice which is unnecessary overhead.

Comment (lasley) [inline app/models/pawn.rb]: Memory is cheaper than CPU

**Comment by lasley** [comment]
Excellent tests. I would still recommend refactoring to use the `:send` like we spoke about in the call, but this is solid coverage & should transfer over easily :+1:

File: app/models/pawn.rb (hunk 1 above)
Comment (rjdellecese) [inline app/models/pawn.rb]: Ooh nice! I'll do that.

**Comment by rjdellecese** [comment]
@lasley I think this should now account for all of your comments. I'd be interested to hear whether you think there's a better way to incorporate the `color_modifier` variable, though!

**Comment by rjdellecese** [comment]
Also, using a modifier variable removed **all** of the Rubocop alerts! Woohoo!

**Comment by lasley** [comment]
Amazing what a little bit of refactoring does to those Rubocop errors huh? There is almost always a way to eliminate them ;)

File: app/models/pawn.rb [hunk 5]
```diff
@@ -1,2 +1,51 @@
 class Pawn < ChessPiece
+  def valid_move?(next_x, next_y)
+    return false unless within_movement_constraints?(next_x, next_y)
+    super
+  end
+
+  private
+
+  def within_movement_constraints?(next_x, next_y)
+    color_modifier = color == 'white' ? 1 : -1
```
Comment (lasley) [inline app/models/pawn.rb]: Implemented perfectly :+1:

**Pull request merged**